### PR TITLE
Allow admin to upload files.

### PR DESCRIPTION
**Pull request by charlesx2013**
Adding carrier wave for both photos and audio.

#81

### CODE REVIEW

File: app/assets/javascripts/authentication/components/TestingPage.js [hunk 1]
```diff
@@ -92,6 +99,34 @@ class TestingPage extends React.Component {
     });
   }
 
+  onImage(value) {
+    console.log('image')
```
Comment (warrenshen) [inline app/assets/javascripts/authentication/components/TestingPage.js]: console

File: app/assets/javascripts/authentication/components/TestingPage.js [hunk 2]
```diff
@@ -92,6 +99,34 @@ class TestingPage extends React.Component {
...
+      },
+    }
+    request.post('/api/admins/subsections/1/components', params, (response) => {
+      this.setState({ image: response.component.photo.image_url })
+      console.log(response.component.photo.image_url)
+    })
+  }
+
+  onAudio(audio) {
+    console.log('audio')
```
Comment (warrenshen) [inline app/assets/javascripts/authentication/components/TestingPage.js]: Same

File: app/assets/javascripts/authentication/components/TestingPage.js [hunk 3]
```diff
@@ -92,6 +99,34 @@ class TestingPage extends React.Component {
...
+    const params = {
+      component: {
+        audio_data: audio,
+      }
+    }
+    request.post('/api/admins/subsections/1/components', params, (response) => {
+      this.setState({ url: response.component.audio_url })
+      console.log(response.component)
```
Comment (warrenshen) [inline app/assets/javascripts/authentication/components/TestingPage.js]: Same

File: app/assets/javascripts/shared/components/forms/AudioUploadInput.js [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    super(props)
+
+    this.state = {
+      audio: '',
+    }
+    this.setChosenAudio = this.setChosenAudio.bind(this)
+  }
+
+  setChosenAudio(audio) {
+    console.log('loaded')
```
Comment (warrenshen) [inline app/assets/javascripts/shared/components/forms/AudioUploadInput.js]: Same

File: app/assets/javascripts/shared/components/forms/ImageUploadInput.js [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+  }
+
+  render() {
+    return (
+      <div>
+        {this.renderImage()}
+        <div>{this.props.label}</div>
+        <UploadInput
+          onChange={this.setChosenImage}
+          accept='image/jpg, image/jpeg, image/png'
```
Comment (warrenshen) [inline app/assets/javascripts/shared/components/forms/ImageUploadInput.js]: I would define this constant somewhere, but I guess you only use it one place.

File: app/assets/javascripts/students/components/UpdateStudentPage.js [hunk 6]
```diff
@@ -96,6 +102,10 @@ class UpdateStudentPage extends React.Component {
     this.setState({ formFields: formFields })
   }
 
+  handleImage(e) {
+    debugger
```
Comment (warrenshen) [inline app/assets/javascripts/students/components/UpdateStudentPage.js]: .

File: app/controllers/api/admins/components_controller.rb [hunk 7]
```diff
@@ -2,10 +2,14 @@ class Api::Admins::ComponentsController < Api::Admins::BaseController
...
   load_and_authorize_resource :component, through: :subsection, shallow: :true
 
+  prepend_before_filter :convert_image, only: [:create, :update]
+  prepend_before_filter :convert_audio, only: [:create, :update]
+
   def create
     if @component.save
       render json: @component, serializer: ComponentSerializer
     else
+      puts @component.errors.messages
```
Comment (warrenshen) [inline app/controllers/api/admins/components_controller.rb]: .

File: app/assets/javascripts/shared/components/forms/UploadInput.js [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+
+class UploadInput extends React.Component {
+  constructor(props) {
+    super(props)
+
+    this.setChosenFile = this.setChosenFile.bind(this)
+  }
+
+  setChosenFile(e) {
+    console.log('chosen1')
```
Comment (warrenshen) [inline app/assets/javascripts/shared/components/forms/UploadInput.js]: ......

File: app/assets/javascripts/authentication/components/TestingPage.js [hunk 9]
```diff
@@ -92,6 +99,34 @@ class TestingPage extends React.Component {
...
+    const params = {
+      component: {
+        photo_attributes: {
+          image_data: value,
+        },
+      },
+    }
+    request.post('/api/admins/subsections/1/components', params, (response) => {
+      this.setState({ image: response.component.photo.image_url })
+      console.log(response.component.photo.image_url)
```
Comment (warrenshen) [inline app/assets/javascripts/authentication/components/TestingPage.js]: Same.

**Comment by charlesx2013** [comment]
👍

**Pull request merged**